File: Sources/custom_adapter/AdmobAdapter/Objc/AlxAdmobBannerAdapter.m
//
//  AlxAdmobBannerAdapter.m
//  AlxAdsOCDemo
//

#import "AlxAdmobBannerAdapter.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"AlxAdmobBannerAdapter";

@interface AlxAdmobBannerAdapter () <AlxBannerViewAdDelegate>

@property (nonatomic, strong, nullable) AlxBannerAdView *bannerAd;
@property (nonatomic, weak, nullable) id<GADMediationBannerAdEventDelegate> delegate;
@property (nonatomic, copy, nullable) GADMediationBannerLoadCompletionHandler completionHandler;

@end

@implementation AlxAdmobBannerAdapter

- (void)loadBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration
                   completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler {
    NSLog(@"%@: loadBanner", TAG);
    
    NSDictionary *params = [AlxAdmobBaseAdapter parseAdparameterFor:adConfiguration.credentials];
    if (!params) {
        NSString *errorStr = @"The parameter field is not found in the adConfiguration object";
        NSLog(@"%@: config params is empty", TAG);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }
    
    if (!AlxAdmobBaseAdapter.isInitialized) {
        [AlxAdmobBaseAdapter initSdkFor:params];
    }
    
    NSString *adId = params[@"unitid"];
    if (!adId || adId.length == 0) {
        NSString *errorStr = @"unitid is empty in the parameter configuration";
        NSLog(@"%@: error: %@", TAG, errorStr);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }
    
    NSLog(@"%@: loadBanner unitid=%@", TAG, adId);
    self.completionHandler = completionHandler;
    
    // Load ad
    CGSize adSize = adConfiguration.adSize.size;
    self.bannerAd = [[AlxBannerAdView alloc] initWithFrame:CGRectMake(0, 0, adSize.width, adSize.height)];
    self.bannerAd.delegate = self;
    self.bannerAd.refreshInterval = 0;
    self.bannerAd.rootViewController = adConfiguration.topViewController;
    [self.bannerAd loadAdWithAdUnitId:adId];
}

#pragma mark - GADMediationBannerAd

- (UIView *)view {
    return self.bannerAd ?: [[UIView alloc] init];
}

#pragma mark - AlxBannerViewAdDelegate

- (void)bannerViewAdLoad:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdLoad", TAG);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(self, nil);
    }
}

- (void)bannerViewAdFailToLoad:(AlxBannerAdView *)bannerView didFailWithError:(NSError *)error {
    NSLog(@"%@: bannerViewAdFailToLoad: %@", TAG, error.localizedDescription);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(self, error);
    }
}

- (void)bannerViewAdImpression:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdImpression", TAG);
    [self.delegate reportImpression];
}

- (void)bannerViewAdClick:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdClick", TAG);
    [self.delegate reportClick];
}

- (void)bannerViewAdClose:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdClose", TAG);
    [self.delegate didDismissFullScreenView];
}

@end

File: custom_adapter/AdmobAdapter/AlxAdmobBannerAdapter.swift
//
//  AlxAdmobBannerAdapter.swift
//

import Foundation
import GoogleMobileAds
import AlxAds

@objc(AlxAdmobBannerAdapter)
public class AlxAdmobBannerAdapter:AlxAdmobBaseAdapter,MediationBannerAd{
    
    private static let TAG = "AlxAdmobBannerAdapter"
    
    // Mark - Banner Ad
    private var bannerAd:AlxBannerAdView? = nil
    // The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
    private var delegate:MediationBannerAdEventDelegate? = nil
    // Completion handler called after ad load
    private var completionHandler: GADMediationBannerLoadCompletionHandler? = nil
    
    public func loadBanner(for adConfiguration: MediationBannerAdConfiguration, completionHandler: @escaping GADMediationBannerLoadCompletionHandler) {
        NSLog("%@: loadBanner",AlxAdmobBannerAdapter.TAG)
        guard let params = AlxAdmobBaseAdapter.parseAdparameter(for: adConfiguration.credentials) else {
            let errorStr="The parameter field is not found in the adConfiguration object"
            NSLog("%@: config params is empty",AlxAdmobBannerAdapter.TAG)
            self.delegate=completionHandler(nil,self.error(code: -100,msg: errorStr))
            return
        }
        
        if !AlxAdmobBaseAdapter.isInitialized {
            AlxAdmobBaseAdapter.initSdk(for: params)
        }
        
        guard let adId = params["unitid"] as? String,!adId.isEmpty else{
            let errorStr="unitid is empty in the parameter configuration"
            NSLog("%@: error: %@",AlxAdmobBannerAdapter.TAG,errorStr)
            self.delegate=completionHandler(nil,self.error(code: -100,msg: errorStr))
            return
        }
        
        NSLog("%@: loadBanner unitid=%@",AlxAdmobBannerAdapter.TAG,adId)
        self.completionHandler = completionHandler
        
        // load ad
        let adSize = CGSize(width: adConfiguration.adSize.size.width, height: adConfiguration.adSize.size.height)
        self.bannerAd=AlxBannerAdView(frame: CGRect(origin: .zero, size: adSize))
        self.bannerAd?.delegate = self
        self.bannerAd?.refreshInterval = 0
        self.bannerAd?.rootViewController=adConfiguration.topViewController
        self.bannerAd?.loadAd(adUnitId: adId)
    }
    
    
    public var view: UIView {
        return bannerAd ?? UIView()
    }
}

extension AlxAdmobBannerAdapter:AlxBannerViewAdDelegate {
    
    
    public func bannerViewAdLoad(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdLoad",AlxAdmobBannerAdapter.TAG)
        if let handler=self.completionHandler{
            self.delegate=handler(self,nil)
        }
    }
    
    public func bannerViewAdFailToLoad(_ bannerView: AlxBannerAdView, didFailWithError error: Error){
        NSLog("%@: bannerViewAdFailToLoad: %@",AlxAdmobBannerAdapter.TAG,error.localizedDescription)
        if let handler=self.completionHandler{
            self.delegate=handler(self,error)
        }
    }
    
    public func bannerViewAdImpression(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdImpression",AlxAdmobBannerAdapter.TAG)
        self.delegate?.reportImpression()
    }
    
    public func bannerViewAdClick(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdClick",AlxAdmobBannerAdapter.TAG)
        self.delegate?.reportClick()
    }
    
    public func bannerViewAdClose(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdClose",AlxAdmobBannerAdapter.TAG)
        self.delegate?.didDismissFullScreenView()
    }
    
    
}
